### PR TITLE
O2 style amp for gpt3 ptuning

### PR DESCRIPTION
**Pull request by JimmyZhang12**
1. Fixes a precision bug for ptuning which improves perf by  ~25% 
- If model is in bf16, Pytorch autocasting currently unnecessarily casts operations to be in fp16 from bf16, and casts the output back again to bf16 
- This fix keeps the precision in bf16, removing many unnecessary casts

2. Implements amp O2 which improves convergence times

- O2 accumulates prompt encoder gradients in fp32 master weights, rather than in bf16 or fp16 currently

Capture
(validation loss of GPT3 530b on 7 nodes, O2 (orange) vs baseline (pink))

### CODE REVIEW

**Review by vadam5 — COMMENTED** [review]
All the prompt learning edits look good to me. Waiting for @titu1994 or someone else outside of the NLP team to review the changes to NeMo core.

**Review by titu1994 — COMMENTED** [review]
@MaximumEntropy  / @okuchaiev for review of optimizer_with_main_params.py. To me, that change looks fine.

No other domain uses that, so impact should be contained to Nemo Megatron.

**Comment by MaximumEntropy** [comment]
Thanks @JimmyZhang12! Can we make similar changes to T5 prompt learning as well?

**Comment by JimmyZhang12** [comment]
> Thanks @JimmyZhang12! Can we make similar changes to T5 prompt learning as well?

Yes, the changes to t5 should be very similar

**Comment by JimmyZhang12** [comment]
@titu1994 @vadam5 Can this be approved?

**Review by vadam5 — APPROVED** [review]
LGTM!

**Review by okuchaiev — APPROVED** [review]
LGTM

**Comment by okuchaiev** [comment]
@MaximumEntropy will this benefits adapters tuning?
@JimmyZhang12 can you please submit similar PR for T5?
